File: easy-random-bean-validation/src/test/java/org/jeasy/random/validation/PastOrPresentAnnotationHandlerTest.java
package org.jeasy.random.validation;

import static org.assertj.core.api.Assertions.assertThat;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import jakarta.validation.constraints.PastOrPresent;
import java.time.*;
import java.time.chrono.HijrahDate;
import java.time.chrono.JapaneseDate;
import java.time.chrono.MinguoDate;
import java.time.chrono.ThaiBuddhistDate;
import java.util.Calendar;
import java.util.Date;
import java.util.Set;
import org.jeasy.random.EasyRandom;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class PastOrPresentAnnotationHandlerTest {

  private EasyRandom easyRandom;

  @BeforeEach
  void setUp() {
    easyRandom = new EasyRandom();
  }

  @Test
  void generatedBeanShouldBeValidAccordingToValidationConstraints() {
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    assertThat(testBean.getTestDate()).isBeforeOrEqualTo(Date.from(Instant.now()));
    assertThat(testBean.getTestCalendar()).isLessThanOrEqualTo(Calendar.getInstance());
    assertThat(testBean.getTestInstant()).isBeforeOrEqualTo(Instant.now());
    assertThat(testBean.getTestLocalDate()).isBeforeOrEqualTo(LocalDate.now());
    assertThat(testBean.getTestLocalDateTime()).isBeforeOrEqualTo(LocalDateTime.now());
    assertThat(testBean.getTestLocalTime()).isBeforeOrEqualTo(LocalTime.now());
    assertThat(testBean.getTestMonthDay()).isLessThanOrEqualTo(MonthDay.now());
    assertThat(testBean.getTestOffsetDateTime()).isBeforeOrEqualTo(OffsetDateTime.now());
    assertThat(testBean.getTestOffsetTime()).isBeforeOrEqualTo(OffsetTime.now());
    assertThat(testBean.getTestYear()).isLessThanOrEqualTo(Year.now());
    assertThat(testBean.getTestYearMonth()).isLessThanOrEqualTo(YearMonth.now());
    assertThat(testBean.getTestZonedDateTime()).isBeforeOrEqualTo(ZonedDateTime.now());
    assertThat(testBean.getTestHijrahDate()).isLessThanOrEqualTo(HijrahDate.now());
    assertThat(testBean.getTestJapaneseDate()).isLessThanOrEqualTo(JapaneseDate.now());
    assertThat(testBean.getTestMinguoDate()).isLessThanOrEqualTo(MinguoDate.now());
    assertThat(testBean.getTestThaiBuddhistDate()).isLessThanOrEqualTo(ThaiBuddhistDate.now());
  }

  @Test
  void generatedBeanShouldBeValidUsingBeanValidationApi() {
    TestBean testBean = easyRandom.nextObject(TestBean.class);

    Validator validator;
    try (ValidatorFactory validatorFactory = Validation.buildDefaultValidatorFactory()) {
      validator = validatorFactory.getValidator();
    }
    Set<ConstraintViolation<TestBean>> violations = validator.validate(testBean);

    assertThat(violations).isEmpty();
  }

  static class TestBean {
    @PastOrPresent private Date testDate;

    public Date getTestDate() {
      return testDate;
    }

    @PastOrPresent private Calendar testCalendar;

    public Calendar getTestCalendar() {
      return testCalendar;
    }

    @PastOrPresent private Instant testInstant;

    public Instant getTestInstant() {
      return testInstant;
    }

    @PastOrPresent private LocalDate testLocalDate;

    public LocalDate getTestLocalDate() {
      return testLocalDate;
    }

    @PastOrPresent private LocalDateTime testLocalDateTime;

    public LocalDateTime getTestLocalDateTime() {
      return testLocalDateTime;
    }

    @PastOrPresent private LocalTime testLocalTime;

    public LocalTime getTestLocalTime() {
      return testLocalTime;
    }

    @PastOrPresent private MonthDay testMonthDay;

    public MonthDay getTestMonthDay() {
      return testMonthDay;
    }

    @PastOrPresent private OffsetDateTime testOffsetDateTime;

    public OffsetDateTime getTestOffsetDateTime() {
      return testOffsetDateTime;
    }

    @PastOrPresent private OffsetTime testOffsetTime;

    public OffsetTime getTestOffsetTime() {
      return testOffsetTime;
    }

    @PastOrPresent private Year testYear;

    public Year getTestYear() {
      return testYear;
    }

    @PastOrPresent private YearMonth testYearMonth;

    public YearMonth getTestYearMonth() {
      return testYearMonth;
    }

    @PastOrPresent private ZonedDateTime testZonedDateTime;

    public ZonedDateTime getTestZonedDateTime() {
      return testZonedDateTime;
    }

    @PastOrPresent private HijrahDate testHijrahDate;

    public HijrahDate getTestHijrahDate() {
      return testHijrahDate;
    }

    @PastOrPresent private JapaneseDate testJapaneseDate;

    public JapaneseDate getTestJapaneseDate() {
      return testJapaneseDate;
    }

    @PastOrPresent private MinguoDate testMinguoDate;

    public MinguoDate getTestMinguoDate() {
      return testMinguoDate;
    }

    @PastOrPresent private ThaiBuddhistDate testThaiBuddhistDate;

    public ThaiBuddhistDate getTestThaiBuddhistDate() {
      return testThaiBuddhistDate;
    }
  }
}

File: bean-validation/src/test/kotlin/io/github/krandom/validation/PastOrPresentAnnotationHandlerTest.kt
package io.github.krandom.validation

import io.github.krandom.KRandom
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.comparables.shouldBeLessThanOrEqualTo
import io.kotest.matchers.nulls.shouldNotBeNull
import jakarta.validation.Validation
import jakarta.validation.Validator
import jakarta.validation.constraints.PastOrPresent
import java.time.Instant
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.MonthDay
import java.time.OffsetDateTime
import java.time.OffsetTime
import java.time.Year
import java.time.YearMonth
import java.time.ZonedDateTime
import java.time.chrono.HijrahDate
import java.time.chrono.JapaneseDate
import java.time.chrono.MinguoDate
import java.time.chrono.ThaiBuddhistDate
import java.util.Calendar
import java.util.Date
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

internal class PastOrPresentAnnotationHandlerTest {
  private lateinit var kRandom: KRandom

  @BeforeEach
  fun setUp() {
    kRandom = KRandom()
  }

  @Test
  fun `generated bean should be valid according to validation constraints`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    testBean.shouldNotBeNull()
    testBean.testDate shouldBeLessThanOrEqualTo Date.from(Instant.now())
    testBean.testCalendar shouldBeLessThanOrEqualTo Calendar.getInstance()
    testBean.testInstant shouldBeLessThanOrEqualTo Instant.now()
    testBean.testLocalDate shouldBeLessThanOrEqualTo LocalDate.now()
    testBean.testLocalDateTime shouldBeLessThanOrEqualTo LocalDateTime.now()
    testBean.testLocalTime shouldBeLessThanOrEqualTo LocalTime.now()
    testBean.testMonthDay shouldBeLessThanOrEqualTo MonthDay.now()
    testBean.testOffsetDateTime shouldBeLessThanOrEqualTo OffsetDateTime.now()
    testBean.testOffsetTime shouldBeLessThanOrEqualTo OffsetTime.now()
    testBean.testYear shouldBeLessThanOrEqualTo Year.now()
    testBean.testYearMonth shouldBeLessThanOrEqualTo YearMonth.now()
    testBean.testZonedDateTime shouldBeLessThanOrEqualTo ZonedDateTime.now()
    testBean.testHijrahDate shouldBeLessThanOrEqualTo HijrahDate.now()
    testBean.testJapaneseDate shouldBeLessThanOrEqualTo JapaneseDate.now()
    testBean.testMinguoDate shouldBeLessThanOrEqualTo MinguoDate.now()
    testBean.testThaiBuddhistDate shouldBeLessThanOrEqualTo ThaiBuddhistDate.now()
  }

  @Test
  fun `generated bean should be valid using bean validation api`() {
    val testBean = kRandom.nextObject(TestBean::class.java)

    val validator: Validator
    Validation.buildDefaultValidatorFactory().use { validatorFactory ->
      validator = validatorFactory.validator
    }
    val violations = validator.validate(testBean)

    violations.shouldBeEmpty()
  }

  internal data class TestBean(
    @field:PastOrPresent val testDate: Date,
    @field:PastOrPresent val testCalendar: Calendar,
    @field:PastOrPresent val testInstant: Instant,
    @field:PastOrPresent val testLocalDate: LocalDate,
    @field:PastOrPresent val testLocalDateTime: LocalDateTime,
    @field:PastOrPresent val testLocalTime: LocalTime,
    @field:PastOrPresent val testMonthDay: MonthDay,
    @field:PastOrPresent val testOffsetDateTime: OffsetDateTime,
    @field:PastOrPresent val testOffsetTime: OffsetTime,
    @field:PastOrPresent val testYear: Year,
    @field:PastOrPresent val testYearMonth: YearMonth,
    @field:PastOrPresent val testZonedDateTime: ZonedDateTime,
    @field:PastOrPresent val testHijrahDate: HijrahDate,
    @field:PastOrPresent val testJapaneseDate: JapaneseDate,
    @field:PastOrPresent val testMinguoDate: MinguoDate,
    @field:PastOrPresent val testThaiBuddhistDate: ThaiBuddhistDate,
  )
}
